models/event.py

```diff
@@ -145,34 +231,106 @@ class Meta:
   objects = SpeedRunManager()
   event = models.ForeignKey('Event', on_delete=models.PROTECT, default=LatestEvent)
   name = models.CharField(max_length=64,editable=False)
-  deprecated_runners = models.CharField(max_length=1024,blank=True,verbose_name='*DEPRECATED* Runners') # This field is now deprecated, we should eventually set up a way to migrate the old set-up to use the donor links
+  deprecated_runners = models.CharField(max_length=1024,blank=True,verbose_name='*DEPRECATED* Runners', editable=False) # This field is now deprecated, we should eventually set up a way to migrate the old set-up to use the donor links
   description = models.TextField(max_length=1024,blank=True)
-  starttime = models.DateTimeField(verbose_name='Start Time')
-  endtime = models.DateTimeField(verbose_name='End Time')
-  runners = models.ManyToManyField('Donor', blank=True, null=True)
+  starttime = models.DateTimeField(verbose_name='Start Time', editable=False, null=True)
+  endtime = models.DateTimeField(verbose_name='End Time', editable=False, null=True)
+  order = models.IntegerField(editable=False, null=True)  # can be temporarily null when moving runs around, or null when they haven't been slotted in yet
+  run_time = TimestampField(always_show_h=True)
+  setup_time = TimestampField(always_show_h=True)
+  runners = models.ManyToManyField('Runner')
   class Meta:
     app_label = 'tracker'
     verbose_name = 'Speed Run'
-    unique_together = ( 'name','event' )
-    ordering = [ 'event__date', 'starttime' ]
+    unique_together = (( 'name','event' ), ('event', 'order'))
+    ordering = [ 'event__date', 'order' ]
+
   def natural_key(self):
     return (self.name,self.event.natural_key())
+
   def clean(self):
-	if not self.name:
-	  raise ValidationError('Name cannot be blank')
+    if not self.name:
+      raise ValidationError('Name cannot be blank')
+
+  def save(self, fix_time=True, fix_runners=True, *args, **kwargs):
+    fix_time = self.order and self.run_time and self.setup_time and fix_time # we can't fix the time without all of these
+    fix_runners = self.id and fix_runners
+    i = TimestampField.time_string_to_int
+    if fix_time:
+      prev = SpeedRun.objects.filter(event=self.event, order__lt=self.order).last()
+      if prev:
+        self.starttime = prev.starttime + datetime.timedelta(milliseconds=i(prev.run_time)+i(prev.setup_time))
+      else:
+        self.starttime = datetime.datetime.combine(self.event.date, datetime.time(12, tzinfo=self.event.timezone))
+      self.endtime = self.starttime + datetime.timedelta(milliseconds=i(self.run_time)+i(self.setup_time))
+    if fix_runners:
+      if not self.runners.exists():
+        try:
+          self.runners.add(*[Runner.objects.get_by_natural_key(r.strip()) for r in self.deprecated_runners.split(',')])
+        except Runner.DoesNotExist:
+          pass
+      if self.runners.exists():
+        self.deprecated_runners = u', '.join(unicode(r) for r in self.runners.all())
```

Comment:
So you're planning on using the 'deprecated_runners' field as a view cache of the list of runners? I guess that's fine, but personally I'd rather migrate them all out using a data migration, and just make methods on SpeedRun for getting and setting the list of runners from the string key. Thoughts?